text/0005-application-formalization.md

```diff
@@ -0,0 +1,191 @@
+- Start Date: 10/09/2023
+- Contributors:
+  - [Craig McNally](cmcnally@ebsco.com)
+  - [Vince Bareau](vbareau@ebsco.com)
+- RFC PRs:
+  - DRAFT REVIEW: https://github.com/folio-org/rfcs/pull/14
+  - PRELIMINARY REVIEW: TBD
+  - PUBLIC REVIEW: TBD
+  - FINAL REVIEW: TBD
+- Outcome: <!-- Leave this blank. Will eventually be either ACCEPTED or REJECTED -->
+
+# Application Formalization
+
+## Summary
+It is proposed that an Application be defined as the minimal but complete set of elements which together are intended to deliver a specific solution to Folio. The intention is to group together all the closely related components which will need to evolve together, but independently of other parts (i.e. other applications) of Folio.
+
+## Motivation
+The existing Folio project lacks a formal definition of an Application structure. A consequence is that Folio is stuck in a mode of monolithic releases which are ever increasing in size.  There have also been repeated complaints that the setup and management of Folio can be difficult due to the number modules involved.
+
+## Scope
+* Application composition
+* Application versioning
+* Application dependencies
+* Application definition/descriptors
+* Application management is ***out of scope***
+* Application-specific UI bundles is ***out of scope***
+* How movement of modules between Applications will work is ***out of scope***
+
+## Detailed Explanation/Design
+
+### Terminology
+* Application:  The minimal but complete set of elements which together are intended to deliver a specific solution to Folio.
+* Bounded context:  The boundary of the microservice.  It "owns" the storage for the service.
+
+### Specifics
+
+#### Application Composition
+Applications can be thought of as vertical slices of functionality.  As such, they're are comprised of a combination of the following:
+* Backend Modules 
+* Edge Modules 
+* UI Modules 
+* UI Plugins 
+
+Applications will typically be made up of multiple components spanning all or several of these categories.  However, that's not a requirement.  In fact, in it's simplest form, it's possible for an application to be comprised of a single module from any of the categories listed above.
+
+An important concept introduced by the formalization of Applications, is that Applications become the delivery mechanism for the modules they contain. Therefore, if a new version of a particular module is required, then it follows that the new version of the entire Application must be installed. It then also follows that a Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.  It is recognized that there must be accommodations to allow refactoring or consolidation of Applications, which could involve moving of particular Modules from one Application to another.
```

Comment:
How is it handled when there are different concepts of what an application is? To be specific - many libraries use inventory and have the concept that the instances in inventory are based on data located in inventory itself and potentially combined with data from source record storage. There is a set of libraries however who see the source of instance data as being outside FOLIO - in their union catalog. They have created a module to load that data into FOLIO. How does the inventory application end up packaged in these cases? Will inventory just be an application all by itself that can be loaded via separate applications? What happens in the early stages of this plan where there are "super applications"?

Reply:
I see this working in a similar fashion to how modules are named today in Folio.  We have mod-inventory, and mod-inventory-storage.  If someone has a different notion of what inventory is/should be, they could introduce a new module, e.g. mod-inventory-union-catalog or simply mod-union-catalog.

To bring this back to applications, let's say that we create an "inventory" application, then later someone introduces a mod-union-catalog.  There are options here:
1. Incorporate it into an existing application (inventory, or otherwise).
2. Create a new union catalog application.

The process for determining the best course of action should probably take the following into consideration:
1. module dependencies between mod-union-catalog and other modules
2. does mod-union-catalog provide an alternative implementation of some or all of the interfaces provided by modules?
3. Is the functionality provided by mod-union-catalog something most libraries will use, or is it something we know only a relatively small percentage will use (like mod-gobi for instance)?  If the latter, it's probably better to create a separate application
4. Are there benefits or drawbacks associated with mod-union-catalog being part of the inventory application?  I'm thinking in terms of:  possibly sharing database a la the bounded context stuff, performance, etc.
5. ??? There are probably other considerations I've missed here.

> What happens in the early stages of this plan where there are "super applications"?

We would likely start with fewer, larger applications (side note: I acknowledge the naming of these could be tricky).  As adjustments are made to modules to have fewer strict interdependencies, we can "carve" smaller applications out of these "super applications".  

To take this to an extreme, you could in theory start with a single application - "app-platform-complete" - which is comprised of every module in a given flower release.  Then start to split that up into more, smaller applications.  Maybe it's app-platform-minimal and app-everything-else.  Those can then be slip up into even more, smaller applications, e.g. app-orders, app-gobi, app-organizations, etc.

> It is recognized that there must be accommodations to allow refactoring or consolidation of Applications, which could involve moving of particular Modules from one Application to another.

The point I'm trying to make with this statement in the RFC is that it would be short-sighted to assume that once applications are defined there will be no need to make adjustments to that.  It's recognized, and expected that modules will need to move from one application to another, applications will need to be split/combined, etc.

Reply:
That said, what is the expectation at this stage of the RFC process wrt this thread?

Reply:
Hi @craigmcnally - Thanks for this explanation. This helps a lot. I'm going to keep track of those considerations you outlined for potential documentation, knowing they may change. I feel like further discussion of this can be handled in upcoming rfcs and/or the tri council working group.